component/common/loki/client/queue_client.go

```diff
@@ -0,0 +1,584 @@
+	}()
+
+	var batchesToFlush []queuedBatch
+
+	for {
+		select {
+		case <-c.quit:
+			return
+
+		case <-maxWaitCheck.C:
+			c.batchesMtx.Lock()
+			// Send all batches whose max wait time has been reached
+			for tenantID, b := range c.batches {
+				if b.age() < c.cfg.BatchWait {
+					continue
+				}
+
+				// add to batches to flush, so we can enqueue them later and release the batches lock
+				// as early as possible
+				batchesToFlush = append(batchesToFlush, queuedBatch{
+					TenantID: tenantID,
+					Batch:    b,
+				})
+
+				// deleting assuming that since the batch expired the wait time, it
+				// hasn't been written for some time
+				delete(c.batches, tenantID)
+			}
+
+			c.batchesMtx.Unlock()
+
+			// enqueue batches that were marked as too old
+			for _, qb := range batchesToFlush {
+				c.sendQueue.enqueue(qb)
+			}
+
+			batchesToFlush = batchesToFlush[:0] // renew slide
+		}
+	}
+}
+
+// enqueuePendingBatches will go over the pending batches, and enqueue them in the send queue. If the context's
+// deadline is exceeded in any enqueue operation, this routine exits.
+func (c *queueClient) enqueuePendingBatches(ctx context.Context) {
+	c.batchesMtx.Lock()
+	defer c.batchesMtx.Unlock()
+
+	for tenantID, batch := range c.batches {
+		if !c.sendQueue.enqueueWithCancel(ctx, queuedBatch{
+			TenantID: tenantID,
+			Batch:    batch,
+		}) {
+			// if enqueue times out due to the context timing out, cancel all
+			return
+		}
+	}
+}
+
+func (c *queueClient) sendBatch(ctx context.Context, tenantID string, batch *batch) {
+	buf, entriesCount, err := batch.encode()
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error encoding batch", "error", err)
+		return
+	}
+	bufBytes := float64(len(buf))
+	c.metrics.encodedBytes.WithLabelValues(c.cfg.URL.Host).Add(bufBytes)
+
+	backoff := backoff.New(c.ctx, c.cfg.BackoffConfig)
+	var status int
+	for {
+		start := time.Now()
+		// send uses `timeout` internally, so `context.Background` is good enough.
+		status, err = c.send(ctx, tenantID, buf)
+
+		c.metrics.requestDuration.WithLabelValues(strconv.Itoa(status), c.cfg.URL.Host).Observe(time.Since(start).Seconds())
+
+		// Immediately drop rate limited batches to avoid HOL blocking for other tenants not experiencing throttling
+		if c.cfg.DropRateLimitedBatches && batchIsRateLimited(status) {
+			level.Warn(c.logger).Log("msg", "dropping batch due to rate limiting applied at ingester")
+			c.metrics.droppedBytes.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonRateLimited).Add(bufBytes)
+			c.metrics.droppedEntries.WithLabelValues(c.cfg.URL.Host, tenantID, ReasonRateLimited).Add(float64(entriesCount))
+			return
+		}
+
+		if err == nil {
+			c.metrics.sentBytes.WithLabelValues(c.cfg.URL.Host).Add(bufBytes)
+			c.metrics.sentEntries.WithLabelValues(c.cfg.URL.Host).Add(float64(entriesCount))
+
+			return
+		}
+
+		// Only retry 429s, 500s and connection-level errors.
+		if status > 0 && !batchIsRateLimited(status) && status/100 != 5 {
```

Comment:
Does loki send a retry after header we can use?

Reply:
Just asked, no header from Loki [thread](https://raintank-corp.slack.com/archives/C9T1FLN9K/p1698243835822239)